cli/src/bin.ts

```diff
@@ -0,0 +1,142 @@
+
+    /**
+     * If the file edits are after this one, increment them
+     * by the length of the new text MINUS the length of the old
+     */
+    const oldTextLength = fileEdit.end - fileEdit.start;
+    const lengthDifference = fileEdit.newText.length - oldTextLength;
+    futureFileEdits.forEach((futureFileEdit) => {
+      if (futureFileEdit.start > fileEdit.end) {
+        futureFileEdit.start += lengthDifference;
+        futureFileEdit.end += lengthDifference;
+      }
+    });
+  }
+  return newText;
+};
+
+const program = new Command();
+
+program.version("0.0.1");
+
+const handleError = (uri: string, e: any) => {
+  if (e?.code === "BABEL_PARSER_SYNTAX_ERROR") {
+    console.log(`${uri} - syntax error, skipping`);
+  }
+};
+
+const writeToFiles = async (uriArray: string[]) => {
+  /**
+   * TODO - implement batching to speed things up
+   * TODO - implement pretty readout
+   */
+  for (const uri of uriArray) {
+    try {
+      const fileContents = await readFile(uri, "utf8");
+      const parseResult = parseMachinesFromFile(fileContents);
+      const event = makeXStateUpdateEvent(
+        uri,
+        parseResult.machines.map((machine) => ({
+          parseResult: machine,
+        })),
+      );
+
+      const fileEdits: FileEdit[] = [];
+
+      for (const machine of parseResult.machines) {
+        let machineIndex = 0;
+        if (machine.ast.definition?.tsTypes?.node) {
+          const relativePath = removeExtension(path.basename(uri));
+          const requiresUpdate = doesTsTypesRequireUpdate({
+            fileText: fileContents,
+            machineIndex,
+            node: machine.ast.definition.tsTypes.node,
+            relativePath,
+          });
+
+          if (requiresUpdate) {
+            fileEdits.push({
+              start: machine.ast.definition.tsTypes.node.start!,
+              end: machine.ast.definition.tsTypes.node.end!,
+              newText: `{} as import("./${relativePath}.typegen").Typegen${machineIndex}`,
+            });
+          }
+          machineIndex++;
+        }
+      }
+
+      if (fileEdits.length > 0) {
+        const newFile = processFileEdits(fileContents, fileEdits);
+
+        await writeFile(uri, newFile);
+      }
+
+      await writeToTypegenFile({
+        filePath: uri,
+        event,
+      });
+      console.log(`${uri} - success`);
+    } catch (e) {
+      handleError(uri, e);
+    }
+  }
+};
+
+program
+  .command("typegen")
+  .description("Generate TypeScript types from XState machines")
+  .argument("<files>", "The files to target, expressed as a glob pattern")
+  .option("-w, --watch", "Run the typegen in watch mode")
+  .action(async (files: string, opts: { watch?: boolean }) => {
+    const allFiles = await fg(files);
+
+    await writeToFiles(allFiles);
+
+    if (opts.watch) {
+      watch(files, {}).on("change", (path) => {
+        if (path.endsWith(".typegen.ts")) {
+          return;
+        }
+        writeToFiles([path]);
+      });
+    }
```

Comment:
There are a couple of small problems with this:
1. the used glob matcher is different between the initial scan and the watcher
2. we might miss changes between the initial scan and setting up the watcher
3. we don't handle newly created files in the watcher
4. we don't clean up the outdated typegen files when a file is being removed (this probably requires more thought and shouldn't be part of this PR but it would be great to create an issue to handle this in the future)
5. there is a potential race condition here because we might receive `change` event before the file in question gets fully updated

How to fix those problems?
ad 1. reuse the pattern matching from chokidar - we can call `getWatched()` on the created watcher, it's not populated synchronously though so we should instantiate the watcher (even when we are not running in the watch mode), wait for the `ready` event and call it from there
ad 2. by using the first suggestion the problem should go away or should be minimized. Hard to tell which one without diving into chokidar's internals but my hope is that we could basically ignore all of the events arriving before the `ready` event. I think that the description of the `ignoreInitial` option implies that it's safe to do so. Alternatively, we should only listen for the `ready` event in the non-watch mode and process events asap when we receive them in the watch mode - this probably would be the best as we wouldn't even have to call `getWatched` in the watch mode. We'd just be notified about "adds" during the initial scan. When I think about it - we'd probably do the same in non-watch mode and just mark the single-pass scan as complete from within the `ready` event.
ad 3. kinda mentioned above - just listen for add events
ad 5. we could use `awaitWriteFinish: true`

There are also race conditions here for our own processing - I think that it would be best to ensure that they are impossible by queuing any new notifications related to the same file while our processing for a given file is already in progress.

Reply:
@Andarist Any chance you could PR a change to this branch? I think I'd understand better by reading your code, and you appear to understand a lot more about chokidar than I do.

Reply:
Happy to do that tomorrow 👍